Set up CI with Azure Pipelines

austindrenski opened the pull request:
Supersedes #636.

roji commented:
Promise to do so ASAP, probably tomorrow night... Will also see about giving management rights to you and @YohDeadfall...

austindrenski commented:
Sounds good!

If management rights are complicated, I think we can get by for now with just the ability to stop, restart, and rerun builds. The good news is that it appears Azure delegates to GitHub for those functions:

![image](https://user-images.githubusercontent.com/21338699/45933014-a5398080-bf53-11e8-88f7-344c01a8a78f.png)

austindrenski commented:
@roji I was finally able to get Azure Pipelines correctly configured for another project (including publishing to MyGet and NuGet), and I've updated this accordingly. At this point, I _think_ that the only difference from the AppVeyor version is that tagged GitHub releases are not handled.

One item to note is that I've reverted to the non-dot-prefixed file name for the time being. It's not a big deal to change it back later, but going with the default makes for one less GUI-specific alteration in the setup. Once the Azure team adds it for implicit detection, I can put through a PR to switch it back.

There are two `TODO` items in `azure-pipelines.yml` that require configuration on your end. There is one awkward hack required after you add new service connections [described here in the docs](https://docs.microsoft.com/en-gb/azure/devops/pipelines/process/resources?view=vsts#troubleshooting-authorization-for-a-yaml-pipeline). 

See below for a link to the results of the current PR built against my account, showing the push to an unstable feed (called npgsql-unstable, but points to a personal feed) and a conditionally skipped push to a stable feed (called npgsql, but again just a placeholder):

https://dev.azure.com/austindrenski/Npgsql.EntityFrameworkCore.PostgreSQL/_build/results?buildId=200&view=logs

roji commented:
Thanks for all this work @austindrenski! It's very welcome. As you've probably noticed, unfortunately I'm still a bit busy with other stuff, but my availability will gradually increase in the coming weeks. Just a bit more patience :)

YohDeadfall commented on azure-pipelines.yml:
```diff
@@ -0,0 +1,70 @@
+name: 3.0.0-ci.$(Date:yyyyMMdd).$(Rev:r) # equal to $(Build.BuildNumber)
```
Is it possible to keep the version number as variable?

austindrenski replied:
IIRC `3.0.0-ci.$(Date:yyyyMMdd).$(Rev:r)` should be available during each job via `$(Build.BuildNumber)`. But (again IIRC), you can't assign/access normal variables at name-definition time.

roji replied:
This seems to mean that for now our config only produces CI packages - I don't see any logic for creating release packages, prerelease packages, etc. etc.

We can always add these later though, after this PR is merged.

YohDeadfall commented on azure-pipelines.yml:
```diff
@@ -0,0 +1,70 @@
+    - dev
+    - master
+    - hotfix/*
+    - release/*
```
Would it work for commits in pull requests?

austindrenski replied:
_Sorry for the delay, just circling back to this today._

Yes, I believe this would include any PRs against those branches. Per [the docs](https://docs.microsoft.com/azure/devops/pipelines/build/ci-build-git?view=azure-devops&tabs=yaml#validate-pull-requests):

> Unless you specify `pr` triggers in your YAML file, pull request builds are automatically enabled for all branches.

austindrenski commented:
Updated, latest build available here: 

https://dev.azure.com/austindrenski/Npgsql.EntityFrameworkCore.PostgreSQL/_build/results?buildId=240

austindrenski commented:
@roji While rebasing, I decided to no longer remove the AppVeyor files, so that we can get this merged soon. Then we can deal with setting up the Azure account (and disabling AppVeyor) later on.

roji commented on azure-pipelines.yml:
```diff
@@ -0,0 +1,81 @@
+name: 3.0.0-ci.$(Date:yyyyMMdd).$(Rev:r) # equal to $(Build.BuildNumber)
+trigger:
```
Any reason to limit the branches built to this list, shouldn't we just build all branches?

roji commented on azure-pipelines.yml:
```diff
@@ -0,0 +1,81 @@
+      publishTestResults: true
+- job: Windows
+  pool:
+    vmImage: vs2017-win2016
```
VS2019 was released - we should probably switch to that... [The docs](https://docs.microsoft.com/en-us/azure/devops/pipelines/agents/hosted?view=azure-devops&tabs=yaml#use-a-microsoft-hosted-agent) still say that it's in preview (it was released a couple of days ago), but we should still use image `windows-2019` unless there's a failure...

roji commented on azure-pipelines.yml:
```diff
@@ -0,0 +1,81 @@
+jobs:
+- job: Linux
+  pool:
+    vmImage: ubuntu-16.04
```
It's pretty disappointing they're still stuck on 16.04... FWIW here's the issue: https://github.com/Microsoft/azure-pipelines-image-generation/issues/506, go vote :)

(but it doesn't seem blocking in any way either)

roji commented on azure-pipelines.yml:
```diff
@@ -0,0 +1,81 @@
+      verbosityPack: 'normal'
+  # TODO: needed because DotNetCoreCLI@2 doesn't support authenticated feeds yet.
+  # TODO: see https://github.com/Microsoft/azure-pipelines-tasks/issues/7160
+  - task: NuGetToolInstaller@0
```
I think that as part of the move to Azure Pipelines, we would also move our two feeds away from myget. This would probably simplify the complexity here (see https://docs.microsoft.com/en-us/azure/devops/pipelines/artifacts/nuget?view=azure-devops&tabs=yaml).

We can handle this a bit later though, after merging this PR.

roji commented:
@austindrenski @YohDeadfall what's the status of this? I think @YohDeadfall still needs to sign off, from my side it seems ready as a first step alongside Appveyor, which we can continue to build upon later. A similar setup needs to be done on the Npgsql side as well.

roji commented:
@austindrenski leaving to your to merge.
